constantine/math/polynomials/polynomials.nim

```diff
@@ -48,70 +57,192 @@ type
-       domain: PolyDomainEval[N, Field],
+  PolyEvalDomain*[N: static int, Field] = object
+    ## Metadata for polynomial in Lagrange basis (evaluation form)
+    ## with generic evaluation points
+
+    domain{.align: 64.}: array[N, Field]     # Evaluation domain for a polynomial in Lagrange basis
+
+    vanishing_deriv_poly_eval{.align: 64.}: PolynomialEval[N, Field]     # A'(X) evaluated on domain
+    vanishing_deriv_poly_eval_inv{.align: 64.}: PolynomialEval[N, Field] # 1/A'(X) evaluated on domain
+
+  PolyEvalLinearDomain*[N: static int, Field] = object
+    ## Metadata for polynomial in Lagrange basis (evaluation form)
+    ## with evaluation points linear in [0, ..., n-1]
+    ##
+    # This allows more efficient polynomial division on the domain.
+    # has we can precompute the inverses 1/(xᵢ-z) with xᵢ,z ∈ [0, ..., n-1]
+    # The first element is 1/0 and unused.
+    dom{.align: 64.}: PolyEvalDomain[N, Field]
+    domain_inverses{.align: 64.}: array[N, Field]
+
+# Polynomials in coefficient form
+# ------------------------------------------------------
+
+func evalPolyAt*[N: static int, Field](
+       r: var Field,
+       poly: PolynomialCoef[N, Field],
+       z: Field) =
+  ## Evaluate a polynomial p at z: r <- p(z)
+  # Implementation using the Horner's method
+  r = poly.coefs[poly.coefs.len-1]
+  for i in countdown(poly.coefs.len-2, 0):
+    r *= z
+    r += poly.coefs[i]
+
+func evalPolyAndDerivAt*[N: static int, Field](
+       r: var Field, rprime: var Field,
+       poly: PolynomialCoef[N, Field],
+       z: Field) =
+  ## Evaluate a polynomial p and its formal derivative p'
+  ## at z:
+  ##  r  <- p(z)
+  ##  r' <- p'(z)
+  # Implementation using the Horner's method
+  r = poly.coefs[poly.coefs.len-1]
+  rprime.setZero()
+  for i in countdown(poly.coefs.len-1, 1):
+    rprime *= z
+    rprime += r
+    r *= z
+    r += poly.coefs[i-1]
+
+func formal_derivative*[N, M: static int, Field](
+       polyprime: var PolynomialCoef[N, Field],
+       poly: PolynomialCoef[M, Field]) =
+  ## Compute P'(x), the formal derivative of the polynomial P(x)
+  ## The derivative of aₙ.xⁿ is n.aₙ.xⁿ⁻¹
+  ## Hence the degree of P'(X) is one less than P(X)
+  static: doAssert N == M-1, "N was " & $N & " and M was " & $M
+
+  # For the part lesser or equal 12, we use addition chains
+  staticFor i, 1, min(13, poly.coefs.len):
+    polyprime.coefs[i-1].prod(poly.coefs[i], i)
+  for i in 13 ..< poly.coefs.len:
+    var degree {.noinit.}: Field
+    degree.fromInt(i)
+    polyprime.coefs[i-1].prod(poly.coefs[i], degree)
+
+# Polynomials in evaluation/Lagrange form
+# ------------------------------------------------------
+
+func areStrictlyIncreasing[Field](a: openArray[Field]): bool {.used.} =
+  ## Check whether points in the domain are ordered
+  if a.len == 0:
+    return false
+
+  var prev = a[0].toBig()
+  for i in 1 ..< a.len:
+    let cur = a[i].toBig()
+    if bool(cur <= prev):
+      return false
+  return true
```

Comment:
This is unused for now, as our KZG and IPA protocols use special domains (roots of unity and linear 0, ..., N-1).